service/controller/resource/nodecontroller/create.go

```diff
@@ -0,0 +1,150 @@
+package nodecontroller
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/provider/v1alpha1"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	"k8s.io/client-go/rest"
+
+	"github.com/giantswarm/kvm-operator/service/controller/internal/nodecontroller"
+	"github.com/giantswarm/kvm-operator/service/controller/key"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.Debugf(ctx, "ensuring controller is running for workload cluster")
+
+	desired, err := r.calculateDesiredController(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if needCreate := r.applyUpdateChange(ctx, desired); needCreate {
+		r.logger.Debugf(ctx, "controller needs to be created")
+		go r.applyCreateChangeAsync(ctx, desired)
+	} else {
+		r.logger.Debugf(ctx, "ensured controller is running")
+	}
+
+	return nil
+}
+
+// applyCreateChangeAsync boots the desired controller and saves it in the controllers map
+// if it boots successfully. It should be run as a goroutine to avoid blocking reconciliation of
+// other resources.
+func (r *Resource) applyCreateChangeAsync(ctx context.Context, desired controllerWithConfig) {
+	r.logger.Debugf(ctx, "DELETEME before boot")
+	desired.Boot(ctx)
+	r.logger.Debugf(ctx, "DELETEME after boot")
+	<-desired.Booted() // Wait for the controller to be booted
```

Comment:
Is there a time limit here? Do we expect this function to always return in a reasonable timeframe?

Reply:
Internally it has a retry back off. If the retry fails a few times, it will exit the whole program with `os.Exit` we don't have to worry about this getting stuck permanently.

Reply:
Looks like it tries 7 times with a 1 second linear backoff, so no more than 7 seconds https://github.com/giantswarm/operatorkit/blob/07b70cc9b37090b1c102bcfa35e78f54edffce5c/pkg/controller/controller.go#L236

Reply:
I see - that's a reasonable timeframe.